##Data description:
Dataset of Germany Ebay for used cars including 50,000 data points

##Data fields:
- dateCrawled - When this ad was first crawled.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

##Introduction
The purpose of this project is to clean the data and analyze used car listings


In [149]:
# Import neccessary libraries and load dataframe
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
%matplotlib inline
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [150]:
# Overview
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

##Overview:
- There are 20 columns in datatset.
- Most of them are strings.
- We have uncompleted data for columns: 'vehicleType','gearbox','model','fuelType','notRepairedDamage'
- Column names use camelcase instead of preferred snakecase.
- Less than 20% of data is missing from each uncompleted columns. That is at an acceptable percentage.

In [151]:
umcompleted = ['vehicleType','gearbox','model','fuelType','notRepairedDamage']

In [152]:
# Change column names for easy interpretation
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

##Finding redundant columns and data incorrectly stored
We will start by looking at df.describe() to find columns not providing much information or those has been stored in incorrect formats

In [153]:
# Further look into data
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [154]:
print(autos['offer_type'].value_counts())
print(autos['seller'].value_counts())
print(autos['price'][0:5])
print(autos['registration_year'].max())

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
privat        49999
gewerblich        1
Name: seller, dtype: int64
0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
9999


- Column 'offer_type' and 'seller' arre redundant as they mostly contain one single value. Let's drop them.
- Column 'price' and 'odometer' has character in them thus are stored as string. We can use str.replace() and str.astype() to convert them
- Column 'registration_year' has max value of 9999 and min of 1000. Let's change it to max value of 2016 and min of 1900

In [155]:
# Convert format of 'price' and 'odometer' column
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].str.replace('$','').astype(int)
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','').astype(int)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [156]:
# Dropping columns 'offer_type' and 'price'
autos = autos.drop(columns=['offer_type', 'seller'])

In [157]:
# Changing value of column 'registration_year'
autos = autos[autos['registration_year'].between(1900, 2016)]

##Cleaning suspicious columns
After the first cleaning, the data seems sane, let's have a further look into numeric columns such as 'odometer_km' and 'price'.
###Numeric columns

In [158]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts().sort_index(ascending=False))

(13,)
count     48028.000000
mean     125544.161739
std       40106.751417
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    31029
125000     4960
100000     2110
90000      1696
80000      1396
70000      1199
60000      1137
50000      1008
40000       801
30000       769
20000       763
10000       249
5000        911
Name: odometer_km, dtype: int64


The data seems to be consistent and well distributed. How about column 'price'?

In [159]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().sort_index(ascending=False))

(2334,)
count    4.802800e+04
mean     9.585252e+03
std      4.843817e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.990000e+03
75%      7.400000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999       1
27322222       1
12345678       1
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             1
50            49
49             3
47             1
45             4
40             6
35             1
30             6
29             1
25             5
20             4
18    

We can see from above that price suddenly went up really high and scarce after 50,000. Also, price below 200 may just be scam so let take a value in between to remove outliers.

In [160]:
autos = autos[autos['price'].between(200, 50000)]

In [161]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().sort_index(ascending=False))

(2116,)
count    45602.000000
mean      5760.396145
std       6801.239702
min        200.000000
25%       1300.000000
50%       3200.000000
75%       7590.000000
max      50000.000000
Name: price, dtype: float64
50000      2
49999      4
49900      5
49500      4
49000      3
48999      2
48900      3
48888      1
48850      1
48700      2
48600      1
48500      3
48490      1
48300      1
48000      4
47997      1
47950      1
47900      3
47800      1
47500      3
47499      1
47000      4
46999      1
46990      1
46911      1
46900      3
46800      1
46500      1
46200      1
46000      2
        ... 
320       11
310        1
300      373
299       53
295        1
290       19
285        1
280       30
277        1
275        7
270        6
269        1
260        5
255        1
251        1
250      284
249       13
248        1
240        2
238        1
235        2
230       12
225        8
222       12
220       32
217        1
215        2
210        1
205        1
200     

###Date columns
There are 5 columns that should represent date values as below:
- 'date_crawled': added by the crawler
- 'last_seen': added by the crawler
- 'ad_created': from the website
- 'registration_month': from the website
- 'registration_year': from the website
We have already formatted the 'registration_year' column, the 'registration_month' seems sane so let's look at the 3 other columns.

In [162]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [163]:
# Caculate the distribution of column 'date_crawled'
print(autos['date_crawled'].value_counts().sort_index())
print(autos['date_crawled'].describe())

2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:21    1
2016-03-05 14:07:26    1
2016-03-05 14:07:40    1
2016-03-05 14:07:45    1
2016-03-05 14:08:00    2
2016-03-05 14:08:05    2
2016-03-05 14:08:27    1
2016-03-05 14:08:42    1
2016-03-05 14:09:02    2
2016-03-05 14:09:05    1
2016-03-05 14:09:20    1
2016-03-05 14:09:22    1
2016-03-05 14:09:38    1
2016-03-05 14:09:46    1
2016-03-05 14:09:57    1
2016-03-05 14:09:58    2
2016-03-05 14:10:18    1
2016-03-05 14:10:20    1
2016-03-05 14:10:46    1
2016-03-05 14:11:03    1
2016-03-05 14:11:05    1
2016-03-05 14:11:14    1
2016-03-05 14:11:15    1
2016-03-05 14:11:25    1
2016-03-05 14:11:40    1
2016-03-05 14:11:42    1
2016-03-05 14:11:56    1
2016-03-05 14:12:16    1
                      ..
2016-04-07 10:06:23    1
2016-04-07 10:25:34    1
2016-04-07 10:36:17    1
2016-04-07 10:36:19    1
2016-04-07 10:36:24    1
2016-04-07 10:36:25    1
2016-04-07 10:36:35    1
2016-04-07 10:36:36    1
2016-04-07 10:36:37    2


In [164]:
print(autos['ad_created'].value_counts().sort_index())
print(autos['ad_created'].describe())

2015-06-11 00:00:00       1
2015-08-10 00:00:00       1
2015-09-09 00:00:00       1
2015-11-10 00:00:00       1
2015-12-05 00:00:00       1
2015-12-30 00:00:00       1
2016-01-03 00:00:00       1
2016-01-07 00:00:00       1
2016-01-10 00:00:00       2
2016-01-13 00:00:00       1
2016-01-14 00:00:00       1
2016-01-16 00:00:00       1
2016-01-27 00:00:00       2
2016-02-01 00:00:00       1
2016-02-02 00:00:00       2
2016-02-05 00:00:00       2
2016-02-07 00:00:00       1
2016-02-08 00:00:00       1
2016-02-09 00:00:00       1
2016-02-11 00:00:00       1
2016-02-12 00:00:00       2
2016-02-14 00:00:00       2
2016-02-16 00:00:00       1
2016-02-17 00:00:00       1
2016-02-18 00:00:00       2
2016-02-19 00:00:00       3
2016-02-20 00:00:00       2
2016-02-21 00:00:00       3
2016-02-22 00:00:00       1
2016-02-23 00:00:00       4
                       ... 
2016-03-09 00:00:00    1522
2016-03-10 00:00:00    1469
2016-03-11 00:00:00    1497
2016-03-12 00:00:00    1670
2016-03-13 00:00:00 

In [165]:
print(autos['last_seen'].value_counts().sort_index())
print(autos['last_seen'].describe())

2016-03-05 14:45:46    1
2016-03-05 14:46:02    1
2016-03-05 14:49:34    1
2016-03-05 15:16:11    1
2016-03-05 15:16:47    1
2016-03-05 15:28:10    1
2016-03-05 15:41:30    1
2016-03-05 15:45:43    1
2016-03-05 15:47:38    1
2016-03-05 15:47:44    1
2016-03-05 16:45:57    1
2016-03-05 16:47:28    1
2016-03-05 17:15:45    1
2016-03-05 17:16:14    1
2016-03-05 17:16:23    1
2016-03-05 17:17:02    1
2016-03-05 17:39:19    1
2016-03-05 17:40:14    1
2016-03-05 17:44:50    1
2016-03-05 17:44:54    1
2016-03-05 17:46:01    1
2016-03-05 18:17:58    1
2016-03-05 18:50:38    1
2016-03-05 19:15:08    1
2016-03-05 19:15:42    1
2016-03-05 19:16:36    1
2016-03-05 19:17:17    1
2016-03-05 19:17:50    1
2016-03-05 19:32:34    1
2016-03-05 19:45:47    1
                      ..
2016-04-07 14:58:09    2
2016-04-07 14:58:10    2
2016-04-07 14:58:12    2
2016-04-07 14:58:13    1
2016-04-07 14:58:14    1
2016-04-07 14:58:17    2
2016-04-07 14:58:18    5
2016-04-07 14:58:20    1
2016-04-07 14:58:21    3


There is no abnormality so let them be for now.
##Data analysis
###Relation between brand, price and mileage
If we were car buyers, our decision relies heavily on car brands, price and their odometer. We will find 10 most popular brand in our dataset and work on finding their average price and mileage

In [166]:
# Print 10 most popular brands in dataset
print(autos['brand'].value_counts()[:10])

volkswagen       9670
bmw              5080
opel             4849
mercedes_benz    4440
audi             3989
ford             3141
renault          2130
peugeot          1371
fiat             1149
seat              832
Name: brand, dtype: int64


In [176]:
# Aggreate data on brand, price and mileage to compare them
brand_price = {}
brand_mil = {}
brands = autos['brand'].value_counts(ascending=False)[:10]
pop_brands = brands.index
for brand in pop_brands:
    price = autos.loc[autos['brand'] == brand,'price']
    brand_price[brand] = round(price.mean())
for brand in pop_brands:
    mil = autos.loc[autos['brand'] == brand,'odometer_km']
    brand_mil[brand] = round(mil.mean())
bp_series = pd.Series(brand_price)
bm_series = pd.Series(brand_mil)
df = pd.DataFrame(bp_series, columns=['mean_price'])
df['mean_mil'] = bm_series
df = df.sort_values(by=['mean_price'])
print(df)

               mean_price  mean_mil
renault            2510.0  128099.0
fiat               2926.0  116950.0
opel               3078.0  129232.0
peugeot            3142.0  126929.0
ford               3750.0  124233.0
seat               4505.0  121605.0
volkswagen         5494.0  128799.0
bmw                8163.0  133051.0
mercedes_benz      8252.0  131592.0
audi               9092.0  129797.0


We observed a distinct price gap between the brands:
- Audi, Mercedes, BMW and Volkswagen are more expensive
- Seat, Ford and Peugeot are in between
- Opel, Fiat and Renault are more affordable
However, more expensive cars come with higher odormeter while affordable cars such as Fiat come with much lower mileage.
###Relation between price and unrepaired damage

###Relation between unrepaired damage and price

In [168]:
# Describe column 'unrepaired_damage'
print(autos['unrepaired_damage'].shape)
print(autos['unrepaired_damage'].value_counts())

(45602,)
nein    33492
ja       4296
Name: unrepaired_damage, dtype: int64


In [175]:
repair_price = {}
repair_bool = autos['unrepaired_damage'].value_counts()
repair = repair_bool.index
for i in repair:
    price = autos.loc[autos['unrepaired_damage'] == i,'price']
    repair_price[i] = round(price.mean())
rp_series = pd.Series(repair_price)
df1 = pd.DataFrame(rp_series, columns=['mean_price'])
print(df1)

      mean_price
ja        2362.0
nein      6769.0


From above, we can conclude that cars needing repair is at a staggering 195% cheaper compared to cars not needing repair

##Conclusion
**for buyers**, used cars with well knowned brand are more expensive but cars with less known brand can be a better bargain
**for sellers**, some minor damage should be catered before advertising for better deal